docs/design.md

```diff
@@ -50,6 +63,328 @@ Default Configuration for Zenoh Sessions:
+* rmw_take_serialized_message
+* rmw_take_serialized_message_with_info
+* rmw_wait
+* rmw_subscription_set_on_new_request_callback
+* rmw_subscription_set_content_filter
+* rmw_subscription_get_content_filter
+* rmw_take_loaned_message
+* rmw_take_loaned_message_with_info
+* rmw_return_loaned_message_from_subscription
+* rmw_subscription_get_network_flow_endpoints
+* rmw_subscription_event_init
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_token
+* zc_sample_payload_rcinc
+* z_declare_subscriber
+* z_undeclare_subscriber
+
+## Service Clients
+
+In ROS 2, services are meant to be used for remote procedure calls that will return fairly quickly.
+`rmw_zenoh_cpp` uses Zenoh queryables to implement ROS 2 services.
+When a client wants to make a request, it uses the rmw API `rmw_send_request`.
+`rmw_zenoh_cpp` then calls the Zenoh `z_get` function to send a query out to the network.
+Assuming there is a service server listening to that queryable, it will receive the request, perform a computation, and return the result.
+The result will then be made available to the client via `rmw_take_response`.
+When a new service client is created, a liveliness token of type `SC` is sent out.
+
+### Related RMW APIs
+
+* rmw_create_client
+* rmw_destroy_client
+* rmw_send_request
+* rwm_take_response
+* rmw_take
+* rmw_take_with_info
+* rmw_take_sequence
+* rmw_take_serialized_message
+* rmw_take_serialized_message_with_info
+* rmw_wait
+* rmw_service_server_is_available
+* rmw_client_set_on_new_response_callback
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_token
+* z_get
+* z_attachment_get
+
+## Service Servers
+
+In ROS 2, services are meant to be used for remote procedure calls that will return fairly quickly.
+`rmw_zenoh_cpp` uses Zenoh queryables to implement ROS 2 services.
+When a ROS 2 node wants to advertise a service to the network, it calls `rmw_create_service`.
+`rmw_zenoh_cpp` uses the `z_declare_queryable` Zenoh API to create that service.
+When a client request comes in, `rmw_take_request` is called to send the query to the user callback, which should perform some computation.
+Once the user callback returns, `rmw_send_response` is called to send the response back to the requester.
+When a new service server is created, a liveliness token of type `SS` is sent out.
+
+### Related RMW APIs
+
+* rmw_create_service
+* rmw_destroy_service
+* rmw_take_request
+* rmw_send_response
+* rmw_take
+* rmw_take_with_info
+* rmw_take_sequence
+* rmw_take_serialized_message
+* rmw_take_serialized_message_with_info
+* rmw_wait
+
+### Related Zenoh-c APIs
+
+* zc_liveliness_declare_token
+* z_attachment_get
+* z_declare_queryable
+* z_undeclare_queryable
+* z_query_value
+* z_query_attachment
+
+## Quality of Service (QoS)
+
+The ROS 2 RMW layer defines quite a few Quality of Service settings that are largely derived from DDS.
+Here is an incomplete list of some of the settings and the values that they can take:
+
+* RELIABILITY
```

Comment:
We should highlight that reliability is something we can only set for subs in Zenoh. If Reliability is RELIABLE and History is KEEP_ALL for pubs, we set Congestion to BLOCK. Else it is DROP.